test/src/org/labkey/test/components/targetedms/PrecursorsWebPart.java

```diff
@@ -0,0 +1,162 @@
+package org.labkey.test.components.targetedms;
+
+import org.labkey.test.Locator;
+import org.labkey.test.WebDriverWrapper;
+import org.labkey.test.components.BodyWebPart;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class PrecursorsWebPart extends BodyWebPart<PrecursorsWebPart.Elements>
+{
+    public static final String DEFAULT_TITLE = "Precursors";
+
+    public PrecursorsWebPart(WebDriver driver)
+    {
+        this(driver, 0);
+    }
+
+    public PrecursorsWebPart(WebDriver driver, int index)
+    {
+        super(driver, DEFAULT_TITLE, index);
+        waitForLoad();
+    }
+
+    public void waitForLoad()
+    {
+        WebDriverWrapper.waitFor(() -> getIntensityTableRowCount() > 4, 10000);
+    }
+
+    public PrecursorsWebPart setSortBy(PrecursorsWebPart.SortBy option)
+    {
+        doAndWaitForElementToRefresh(() -> getWrapper().selectOptionByText(elementCache().sortBy, option.toString()),
+                getVisibleTable().append("//tbody/tr[1]/td[1]"), //Checking for staleness in the first element of the table
+                1000);
+        return this;
+    }
+
+    public Locator.XPathLocator getVisibleTable()
+    {
+        if (elementCache().intensityTable.findElement(this).getAttribute("style").contains("display: none;"))
+            return elementCache().ratioTable;
+
+        return elementCache().intensityTable;
+    }
+
+    public List<String> getSortByOptions()
+    {
+        return getWrapper().getSelectOptions(elementCache().sortBy);
+    }
+
+    public PrecursorsWebPart selectIntensity()
+    {
+        getWrapper().checkRadioButton(elementCache().intensity);
+        return this;
+    }
+
+    public PrecursorsWebPart selectRatio()
+    {
+        getWrapper().checkRadioButton(elementCache().ratio);
+        return this;
+    }
+
+    public boolean isSelected(String label)
+    {
+        if (label.equals("Intensity"))
+            return getWrapper().isChecked(elementCache().intensity);
+        else
+            return getWrapper().isChecked(elementCache().ratio);
+    }
+
+    public PrecursorsWebPart selectValueType(String value)
+    {
+        getWrapper()._ext4Helper.selectComboBoxItem(elementCache().valueType, value);
+        return this;
+    }
+
+    public int getIntensityTableRowCount()
+    {
+        return elementCache().intensityTable.append("//tbody/tr").findElements(this).size();
+    }
+
+    public int getRatioTableRowCount()
+    {
+        return elementCache().ratioTable.append("//tbody/tr").findElements(this).size();
+    }
+
+    public String getIntensityTableElement(int row, int column)
```

Comment:
may want to be careful about using the term "Element". I have seen other test libraries where element refers to the actual web element. Maybe replace it with "Value"?